runtime/jcl/common/shared.c

```diff
@@ -546,50 +545,78 @@ getPathProtocolFromURL(JNIEnv* env, jobject url, jmethodID URLgetPathID, jmethod
+getCpeTypeForProtocol(JNIEnv *env, const char *protocol, jsize protocolLen, const char *urlPathChars, jsize urlPathLen, char **correctedPathPtr)
 {
+	J9JavaVM *vm = ((J9VMThread *)env)->javaVM;
+	const char *pathChars = NULL;
+	jsize pathLen = 0;
+
 	Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_Entry();
 
-	if (!protocol) {
-		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_Exit0();
-		return 0;
+	if (NULL == protocol) {
+		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitFail1();
+		return CPE_TYPE_UNKNOWN;
+	}
+	if (!correctURLPath(env, urlPathChars, urlPathLen, correctedPathPtr, &(vm->sharedClassConfig->jclStringFarm))) {
+		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitFail2();
+		return CPE_TYPE_UNKNOWN;
 	}
-	if (strncmp(protocol, "jar", 4)==0) {
+	if (0 == strncmp(protocol, "jar", 4)) {
 		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitJAR();
 		return CPE_TYPE_JAR;
 	}
-	if (strncmp(protocol, "file", 5)==0) {
-		char* endsWith = (char*)(pathChars + (pathLen-4));
-		if ((strncmp(endsWith, ".jar", 4)==0) || (strncmp(endsWith, ".zip", 4)==0) || strstr(pathChars,"!/") || strstr(pathChars,"!\\")) {
+	pathChars = *correctedPathPtr;
+	pathLen = (jsize)strlen(pathChars);
+	if (0 == strncmp(protocol, "file", 5)) {
+		if ((NULL != strstr(pathChars, "!/")) || (NULL != strstr(pathChars, "!\\"))) {
+			/* file is a fat jar */
 			Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitJAR();
 			return CPE_TYPE_JAR;
+		} else if (isPathTypeJimage(pathChars, pathLen)) {
+			Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitJIMAGE();
+			return CPE_TYPE_JIMAGE;
 		} else {
-			char JimageEndsWith[] = DIR_SEPARATOR_STR "lib" DIR_SEPARATOR_STR "modules";
-			IDATA len = LITERAL_STRLEN(JimageEndsWith);
-
-			if (pathLen >= len) {
-				endsWith = (char*)(pathChars + (pathLen - len));
-				if (strncmp(endsWith, JimageEndsWith, len) == 0) {
-					Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitJIMAGE();
-					return CPE_TYPE_JIMAGE;
-				}
+			PORT_ACCESS_FROM_JAVAVM(vm);
+			I_32 result = j9file_attr(pathChars);
+			if (EsIsFile == result) {
+				Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitJAR();
+				return CPE_TYPE_JAR;
+			} else if (EsIsDir == result) {
+				Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitDIR();
+				return CPE_TYPE_DIRECTORY;
+			} else {
+				Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitFail3(result);
+				return CPE_TYPE_UNKNOWN;
 			}
 		}
-		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitDIR();
-		return CPE_TYPE_DIRECTORY;
+	} else {
+		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_UnknownProtocol(protocolLen, protocol, pathLen, pathChars);
+		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitUnknown();
+		return CPE_TYPE_UNKNOWN;
 	}
-	Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_UnknownProtocol(protocolLen, protocol, pathLen, pathChars);
-	Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitUnknown();
-	return CPE_TYPE_UNKNOWN;
```

Comment:
Please revert the addition of the `else` block; not all compilers recognize that the function always has a return value expressed this way.

Reply:
I have removed it.